## This is the tag along code for the stage A quiz

### Author details
#### Name: Jimoh Abdulsomad Abiola
#### Hamoye ID:  15e47f79f981f000

#### Importing the necessary modules

In [87]:
import numpy as np
import pandas as pd

In [88]:
file = pd.read_csv("FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding="latin-1")
file

Area Code      Area  Item Code           Item  Element Code  \
0              4   Algeria       2501     Population           511   
1              4   Algeria       2501     Population          5301   
2              4   Algeria       2901    Grand Total           664   
3              4   Algeria       2901    Grand Total           674   
4              4   Algeria       2901    Grand Total           684   
...          ...       ...        ...            ...           ...   
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit     Y2014  \
0               Total Population - Both sexes     1000 persons  38924.00   
1                    Domestic supply quantity      1000 tonnes      0.00   
2               Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3      Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4          Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                       ...              ...       ...   
60938                                    Food      1000 tonnes     42.00   
60939     Food supply quantity (kg/capita/yr)               kg      3.06   
60940           Food supply (kcal/capita/day)  kcal/capita/day      3.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day      0.10   
60942      Fat supply quantity (g/capita/day)     g/capita/day      0.04   

          Y2015     Y2016     Y2017     Y2018  
0      39728.00  40551.00  41389.00  42228.00  
1          0.00      0.00      0.00      0.00  
2       3379.00   3372.00   3341.00   3322.00  
3         94.35     94.72     92.82     91.83  
4         79.36     77.40     80.19     77.28  
...         ...       ...       ...       ...  
60938     46.00     33.00     19.00     16.00  
60939      3.33      2.35      1.33      1.08  
60940      4.00      3.00      1.00      1.00  
60941      0.11      0.08      0.04      0.04  
60942      0.05      0.03      0.02      0.01  

[60943 rows x 12 columns]

### Deriving insights from the dataset

In [89]:
file.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [90]:
file.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB


In [91]:
file.describe()

Area Code     Item Code  Element Code          Y2014          Y2015  \
count  60943.000000  60943.000000  60943.000000   59354.000000   59395.000000   
mean     134.265576   2687.176706   3814.856456     134.196282     135.235966   
std       72.605709    146.055739   2212.007033    1567.663696    1603.403984   
min        4.000000   2501.000000    511.000000   -1796.000000   -3161.000000   
25%       74.000000   2562.000000    684.000000       0.000000       0.000000   
50%      136.000000   2630.000000   5142.000000       0.090000       0.080000   
75%      195.000000   2775.000000   5511.000000       8.340000       8.460000   
max      276.000000   2961.000000   5911.000000  176405.000000  181137.000000   

               Y2016          Y2017          Y2018  
count   59408.000000   59437.000000   59507.000000  
mean      136.555222     140.917765     143.758381  
std      1640.007194    1671.862359    1710.782658  
min     -3225.000000   -1582.000000   -3396.000000  
25%         0.000000       0.000000       0.000000  
50%         0.080000       0.100000       0.070000  
75%         8.430000       9.000000       9.000000  
max    185960.000000  190873.000000  195875.000000

### Observing 5 random samples from the dataset

In [92]:
samples = file.sample(n=5)
samples

Area Code     Area  Item Code                     Item  Element Code  \
38695        143  Morocco       2948  Milk - Excluding Butter           674   
42423        158    Niger       2618  Pineapples and products           664   
32469        130   Malawi       2745                    Honey          5301   
53880        217     Togo       2899            Miscellaneous           645   
387            4  Algeria       2556  Groundnuts (Shelled Eq)          5142   

                                      Element             Unit  Y2014  Y2015  \
38695  Protein supply quantity (g/capita/day)     g/capita/day   4.82   4.88   
42423           Food supply (kcal/capita/day)  kcal/capita/day   0.00   0.00   
32469                Domestic supply quantity      1000 tonnes   0.00   0.00   
53880     Food supply quantity (kg/capita/yr)               kg   3.65   3.65   
387                                      Food      1000 tonnes   7.00   7.00   

       Y2016  Y2017  Y2018  
38695   4.95   5.06   5.14  
42423   0.00   0.00   0.00  
32469   0.00   0.00   0.00  
53880   3.65   3.65   3.65  
387     6.00   4.00  11.00

# Questions and Solutions

### 1. Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

In [93]:
p_2017=file[file["Element"]=="Processing"].Y2017.sum()
p_2017

292836.0

### 2. Which of the following dataframe methods can be used to access elements across rows and columns?

answer: both .loc() and iloc() method can be used to access elements

### 3. What is the total sum of Wine produced in 2015 and 2018 respectively?

In [94]:
wine= file[file["Item"]=="Wine"]
wine_2015=wine["Y2015"].sum()
wine_2018=wine["Y2018"].sum()
#to two decimal points as in the options
round(wine_2015,2), round(wine_2018,2)

(4251.81, 4039.32)

### 4. Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [95]:
file.groupby("Area")["Y2017"].sum().sort_values(ascending=True)

Area
Sudan (former)                       0.00
Ethiopia PDR                         0.00
Comoros                             59.84
Seychelles                         442.34
Sao Tome and Principe            12662.63
Cabo Verde                       14650.74
Guinea-Bissau                    19102.77
Lesotho                          21267.96
Botswana                         22101.30
Djibouti                         22729.91
Gambia                           23154.18
Gabon                            27979.64
Liberia                          29342.20
Namibia                          29874.89
Central African Republic         29937.00
Congo                            41181.68
Togo                             49841.88
Mauritius                        51114.83
Eswatini                         54343.33
Sierra Leone                     55311.33
Chad                             71594.68
Rwanda                           73663.69
Zimbabwe                         75919.34
Senegal                      

It can be seen from the series above that after sorting in ascending order, Guinea-Bisau is the 7th

### 5. What is the ouput for, S = [['him', 'sell'], [90, 28, 43]], if S[0][1][1] is indexed


In [96]:
S = [['him', 'sell'], [90, 28, 43]]
S[0][1][1]

'e'

The output in the above list indexing is "e"

### 6. What is the total Protein supply quantity in Madagascar in 2015?

In [97]:
mdg= file[file["Area"]=="Madagascar"]
mdg_protein = mdg[mdg["Element"]=="Protein supply quantity (g/capita/day)"]
mdg_protein_2015 = mdg_protein.Y2015
sum_mdg=mdg_protein_2015.sum()
#to 2 decimal places
round(sum_mdg,2)

173.05

### 7. Which of the following is a python inbuilt module?

The math module is the only builtin python module in the options

### 8. What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [98]:
#The null value in Y2014
null_2014 = file['Y2014'].isnull().sum()
null_all = file.isnull().sum()
percent_null = (null_2014/len(file["Y2014"]))*100
#print null_2014 percent_null to 3 decimal places
null_2014,round(percent_null,3)

(1589, 2.607)

It can be seen above that the total null value in the year 2014 is 1589
and the percentage of that to the total data in Y2014, to 2 dp is 2.607%

### 9. What year has the least correlation with "Element Code"?

In [99]:
file.corr()

C:\Users\USER\AppData\Local\Temp\ipykernel_9948\1007512219.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  file.corr()


Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

For the various years 2014,2015,2016,2017, and 2018, there is a corresponding correlation value of 0.024457,0.023889,0.023444,0.02454 and 0.024279

It can be seen that Year 2016 has the least correlation with Element Code 

### 10. y = [(2, 4), (7, 8), (1, 5, 9)]
what indexing will output 8
* 8 can  be found in row index 1
* and in the row index 1, it can be found at the last of the tuple, which can be indexed as either 1 or -1
* it can be indexed as y[1][1] or y[1][-1]
* Only y[1][-1] can be found in the options

I will try that out below

In [100]:
y = [(2, 4), (7, 8), (1, 5, 9)]
y[1][-1]

8

### 11. Which of these python structures is unorderly

Answer is Set

### 12. Given the following python code, what would the output of the code give?

Typeerror will be raised, because tuples are not mutable and therefore do not support item assignment 

I will try it below

In [101]:
my_tuppy = (1,2,5,8)
my_tuppy[2]=6

TypeError: 'tuple' object does not support item assignment

### 13. DataFrame and table included in question 

pd.DataFrame(lst, columns =col, index =[1,2,3]) is the right option

lst and col were included in the question

### 14. What is the total number of unique countries in the dataset

In [102]:
unique_Area=file.Area.unique()
#Total unique_Area
len(unique_Area)

49

There are 49 unique countries in the dataset

### 15. Given a numpy array, how would you select texts that was given in italics

Those numbers were 89,63,92,48 in certain indices

In [103]:
arr = np.array([[94,89,63],[93,92,48],[92,94,56]])
#row 0 and 1 (0:2 or :2), col 1 and 2 (1:3 or 1:)
# from the options, the most suitable is arr[:2,1:]

arr[:2,1:]


array([[89, 63],
       [92, 48]])

The above output array matches with the required

### 16. What is the mean and standard deviation accross the whole dataset for the year 2017 to 2 decimal places

In [104]:
year_2017 = file.Y2017
mean = year_2017.mean()
std = year_2017.std()

#printing mean and std to 2 dp
round(mean,2),round(std,2)

(140.92, 1671.86)

The mean and standard deviation for the year 2017 accross the dataset is 140.92 and 1671.86 to two decimal places respectively  

### 17. What year has the highest stock variation

In [105]:
stock = file[file["Element"]=="Stock Variation"]
year_stock=stock[["Y2014","Y2015","Y2016","Y2017","Y2018"]].sum()

#printing the year with highest sum of Stock Variation and the sum for that year
year_stock.idxmax(),year_stock.max()

('Y2014', 58749.83)

It can be seen above that the year 2014 has the highest sum of Stock variation, with a sum of 58749.83 

### 18. How would you check for the number of rows and columns in a pandas DataFrame named df


### Answer
pd.DataFrame.shape is used to check the number of rows and columns in a DataFrame
In the question, df.shape is used

I will try it out with our dataset named "file"

In [106]:
file.shape

(60943, 12)

Our dataset named "file" has a 60,943 rows and 12 colomns

### 19. Which of the Area had the highest sum in 2017


In [107]:
y_2017 = file.groupby("Area")["Y2017"].sum()
#Year with maximum sum and the sum
y_2017.idxmax(), y_2017.max()

('Nigeria', 1483268.23)

Nigeria has the highest sum in year 2017, with total sum of 1483268.23

### 20. A pandas DataFrame with dimensions (100,3), has how many features and observations 

feature is the number of columns and observation is the number of rows

The shape of a DataFrame is a tuple in the form of (row,column)

Therefore the DataFrame has 3 features(columns) and 100 Obsevations(rows)